model/fn-execution/src/main/proto/beam_fn_api.proto

```diff
@@ -280,11 +280,44 @@ message Metrics {
 
   // User defined metrics
   message User {
-    // TODO: Define it.
+
+    // A key for identifying a metric at the most granular level.
+    message MetricKey {
+      // The step, if any, this metric is associated with.
+      string step = 1;
+
+      // (Required): The namespace of this metric.
+      string namespace = 2;
+
+      // (Required): The name of this metric.
+      string name = 3;
+    }
+
+    // Data associated with a counter metric.
+    message CounterData {
```

Comment:
I only just thought of it while trying to find the JIRA associated with Fn API metrics, but I don't think user-defined metrics should necessarily be conflated with progress reporting. For some contexts, a runner may just support attempted metrics via a proxy over dropwizard. It might be nice to be able to stand this up as an independent service that runners can share.

Reply:
I'm not sure what your trying to get at here with an independent service?

If you are saying that the Metrics type should be defined outside of beam_fn_api.proto in some place that is shared between job management and job execution then I agree.

Reply:
I mean that the message sent from SDK harness to runner should be over an independent channel from progress reporting.

Reply:
The context is that there has been an ask to be able to publish metrics to a metrics pusher that is either independent of the runner (can only support attempted metrics) or supported by a runner. In either case both APIs can share this metrics type or the metrics pusher API can define a new API which is relevant for its use case.